DAOS-16477 mgmt: return suspect engines for pool healthy query

wangshilong opened the pull request:
After significant failures, the system may leave behind some suspect engines that were marked as DEAD by the SWIM protocol, but were not excluded from the system to prevent data loss. An administrator can bring these ranks back online by restarting them.

This PR aims to provide an administrative interface for querying suspect engines following a massive failure. These suspect engines can be retrieved using the daos/dmg --health-only command.

An example of output of dmg pool query --health-only:

Pool 6f450a68-8c7d-4da9-8900-02691650f6a2, ntarget=8, disabled=2, leader=3, version=4, state=Degraded Pool health info:
- Disabled ranks: 1
- Suspect ranks: 2
- Rebuild busy, 0 objs, 0 recs

Features: DmgPoolQueryRanks
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

wangshilong commented:
To reviewers: This PR landed before but got reverted because of conflicts with MD-on-SSD phase2 PR, i refreshed the PR with master and removed a walkaround in the pool query tests(since rebuild bug fixed)

liw reviewed (approved):
C changes look good to me.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15458/1/execution/node/1478/log

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -495,6 +494,7 @@ def __init__(self, base_namespace, index, provider=None, max_storage_tiers=MAX_S
             "ABT_ENV_MAX_NUM_XSTREAMS=100",
             "ABT_MAX_NUM_XSTREAMS=100",
             "DAOS_MD_CAP=1024",
+            "DAOS_POOL_RF=4",
```
By moving this into `default_env_vars` the `DAOS_POOL_RF=4` will only be set for test that do NOT define a `/run/server_config/engines/x/env_vars` in their test yaml, or redfine it in that same yaml setting.  Just confirming that this is the intent.  For example, a test like, `src/tests/ftest/daos_test/rebuild.py` (because its yaml file defines specific `env_vars` entries - overriding the default) will no longer run with `DAOS_POOL_RF=4` with this change. There are currently ~50 functional tests that override the server config env_vars.

wangshilong replied:
@phender I am bit confused with current behavior. for this specific test, i want to reset DAOS_POOL_RF=2, but for all other tests, i want it to be default.

put DAOS_POOL_RF=4 in default_env_vars not match what i need, but don't move it, set it in the server yaml file won't change the value. would you point out how we should fix this? I asked this before...

phender replied:
Which test do you want to use DAOS_POOL_RF=2?  If it is src/tests/ftest/control/dmg_pool_query_ranks.py, then we just need to include that entry in the test yaml (plus any other default values you'd like to retain).  Since the original src/tests/ftest/control/dmg_pool_query_ranks.yaml already defines a server_config/engines/x/env_vars entry it is already overriding the defaults that the test harness would otherwise include in the server yaml it generates. So we'd just need to add DAOS_POOL_RF=2:
```
server_config:
  name: daos_server
  crt_timeout: 5
  engines_per_host: 1
  engines:
    0:
      targets: 4
      nr_xs_helpers: 0
      env_vars:
        - SWIM_SUSPECT_TIMEOUT=10000
        - DAOS_POOL_RF=1
        - DD_MASK=io,epc,rebuild
        - DAOS_POOL_RF=2
```

Then the changes to src/tests/ftest/daos_test/suite.yaml and src/tests/ftest/util/server_utils_params.py can be reverted.

wangshilong replied:
I would probably prefer second method to avoid touching too many files.

phender replied:
I've pushed an update that implements the second method.

phender reviewed (changes requested):
There is at least one fix needed.  I also have a couple questions about the intent of some changes.

wangshilong commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -52,14 +52,21 @@ def test_pool_query_ranks_basic(self):
         self.assertListEqual(
             data['response'].get('disabled_ranks'), [],
-            "Invalid disabled_ranks field: want=[], got={}".format(
+            "Invalid suspect_ranks field: want=[], got={}".format(
```
Yes, should be disabled ranks..

wangshilong commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -80,58 +87,65 @@ def test_pool_query_ranks_mgmt(self):
-            self.pool.wait_for_rebuild_to_end()
+        exclude_rank = all_ranks[0]
+        suspect_rank = all_ranks[1]
+        self.log.info("Starting excluding rank:%d all_ranks=%s", exclude_rank, all_ranks)
```
> The new behavior of only stopping two ranks does not align with the test description or how it currently runs. The test starts 3 ranks (with the option to expand this by changing the `test_servers` count in the test yaml). The old code would stop all three ranks (or more if more servers were specified). Is this fully intentional?

I think test logs need be updated. We don't need stop all three ranks. test intention is to test enable, disable, suspect ranks change accordingly..




> 
> Since we're modifying this test it would be beneficial use the new log step feature. This is helpful for determining how long parts of the test take and provides a unique search string to jump to in the log:

Yup, that is good idea indeed.

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15458/2/testReport/

wangshilong commented:
@phender would you help fix CI env issue? thanks!

daltonbohning commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -433,16 +434,17 @@ class EngineYamlParameters(YamlParameters):
-            "CRT_EVENT_DELAY=1",
-            "COVFILE=/tmp/test.cov"],
+            EnvironmentalVariable("D_LOG_FILE_APPEND_PID", "1"),
+            EnvironmentalVariable("DAOS_POOL_RF", "4", True),
```
Why is this necessary? Couldn't we just let all of these be overridden and then we don't need the new class `EnvironmentalVariable` at all?

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15458/3/execution/node/1150/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15458/4/execution/node/1203/log

wangshilong commented:
@phender  It looks something is broken, would you please check?

`EnvironmentalVariable' object has no attribute 'split'
Stacktrace
Traceback (most recent call last):
  File "/localhome/jenkins/venv/lib64/python3.6/site-packages/avocado/core/test.py", line 747, in _run_avocado
    self.setUp()
  File "/usr/lib/daos/TESTING/ftest/util/telemetry_test_base.py", line 23, in setUp
    super().setUp()
  File "/usr/lib/daos/TESTING/ftest/util/apricot/apricot/test.py", line 813, in setUp
    force_agent_start = self.start_servers()
  File "/usr/lib/daos/TESTING/ftest/util/apricot/apricot/test.py", line 917, in start_servers
    self.setup_servers(server_groups)
  File "/usr/lib/daos/TESTING/ftest/util/apricot/apricot/test.py", line 1044, in setup_servers
    info["access_points"])
  File "/usr/lib/daos/TESTING/ftest/util/apricot/apricot/test.py", line 1192, in configure_manager
    manager.get_params(self)
  File "/usr/lib/daos/TESTING/ftest/util/server_utils.py", line 190, in get_params
    super().get_params(test)
  File "/usr/lib/daos/TESTING/ftest/util/command_utils.py", line 1220, in get_params
    self.manager.job.get_params(test)
  File "/usr/lib/daos/TESTING/ftest/util/server_utils_base.py", line 119, in get_params
    super().get_params(test)
  File "/usr/lib/daos/TESTING/ftest/util/command_utils.py", line 927, in get_params
    self.yaml.get_params(test)
  File "/usr/lib/daos/TESTING/ftest/util/server_utils_params.py", line 184, in get_params
    self.engine_params[-1].get_params(test)
  File "/usr/lib/daos/TESTING/ftest/util/server_utils_params.py", line 542, in get_params
    env_var_dict = {env.split("=")[0]: env.split("=")[1] for env in self.env_vars.value}
  File "/usr/lib/daos/TESTING/ftest/util/server_utils_params.py", line 542, in <dictcomp>
    env_var_dict = {env.split("=")[0]: env.split("=")[1] for env in self.env_vars.value}
AttributeError: 'EnvironmentalVariable' object has no attribute 'split'
`

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15458/5/testReport/

daltonbohning commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -85,10 +85,12 @@ def test_pool_query_ranks_mgmt(self):
         self._verify_ranks(enabled_ranks, data, "enabled_ranks")
         self._verify_ranks(disabled_ranks, data, "disabled_ranks")
 
+        self.log_step(f"Waiting for rebuild to start after excluding pool rank {exclude_rank}")
+        self.pool.wait_for_rebuild_to_start()
```
I thought @wangshilong intentionally removed this before?

wangshilong replied:
No, this is fine.

daltonbohning reviewed (commented):
ftest LGTM

kjacque commented on src/mgmt/tests/srv_drpc_tests.c:
```diff
@@ -1424,8 +1425,9 @@ test_drpc_pool_query_success(void **state)
-	assert_int_equal(ds_mgmt_pool_query_info_in.pi_bits,
-			 DEFAULT_QUERY_BITS | DPI_ENGINES_ENABLED | DPI_ENGINES_DISABLED);
+	assert_non_null(ds_mgmt_pool_query_suspect_ranks_out);
+	flags |= DEFAULT_QUERY_BITS;
+	assert_int_equal(ds_mgmt_pool_query_info_in.pi_bits, DEFAULT_QUERY_BITS | flags);
```
Redundant bitwise OR

```suggestion
	assert_int_equal(ds_mgmt_pool_query_info_in.pi_bits, DEFAULT_QUERY_BITS | flags);
```

kjacque reviewed (approved):
C and Go changes LGTM. Comment is minor.
